src/middleware/jwtAuthentication.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+import jwt from 'jsonwebtoken'
+import logger from 'winston'
+
+import * as client from '../model/clients'
+import * as configIndex from '../config'
+import fs from 'fs'
+import path from 'path'
+
+const TOKEN_PATTERN = /^ *(?:[Bb][Ee][Aa][Rr][Ee][Rr]) +([A-Za-z0-9\-._~+/]+=*) *$/
+
+async function authenticateClient(clientID) {
+  return client.ClientModel.findOne({ clientID }).then((client) => {
+    if (!client) {
+      throw new Error('Client does not exist')
+    }
+    return client
+  })
+}
+
+function resolveJwtSecretOrPublicKey() {
+  let secretOrPublicKey = configIndex.config.get(
+    'authentication:jwt:secretOrPublicKey'
+  )
+
+  try {
+    const publicKeyFilePath = path.resolve(
+      __dirname,
+      '..',
+      '..',
+      'resources',
+      'certs',
+      'jwt',
+      secretOrPublicKey
+    )
+
+    // Check file exists
+    if (fs.existsSync(publicKeyFilePath)) {
+      secretOrPublicKey = fs.readFileSync(publicKeyFilePath).toString()
+    }
```

Comment:
Could we not just make use of the `readFileSync` method to get the file and if nothing comes back (or error thrown) the catch will handle that. Currently we are interacting with the File System twice on the exists, and then on the read

Reply:
@BMartinos Maybe we shouldn't be reading from the file system every time at all.
Could we keep the secret in an in-memory cache? The first time the secret is needed we read it - after that, we just check the cached value

Reply:
Cool, yeah i think thats a better idea. Then we reduce the amount of IO interactions to the file system and speeds up the authentication. Maybe this should then be executed when the server starts, and not have to wait for the first request to come through